File: erlang/pe003.erl
-module(pe003).
-export([start/0]).

maxv() -> 600851475143.

start() ->
  pe003(maxv()).
  
pe003(Number) when (Number =:= 2) ->
  io:format("~w~n", [2]);
    
pe003(Number) when (Number rem 2 =:= 0) ->
  pe003(Number div 2);
    
pe003(Number) ->
  loop003(3, Number).

loop003(Start, Number) when (Number =:= 1) ->
  io:format("~w~n", [Start - 2]);

loop003(Start, Number) when (Number rem Start =:= 0) ->
  loop003(Start + 2, Number div Start);

loop003(Start, Number) ->
  loop003(Start + 2, Number).

File: ocaml/pe003.ml
let divisible x y  = 
  Int64.(rem x y = zero) ;;

let rec pe003 n i =
(*  Printf.printf "n is %Lu and i is %Lu\n" n i ;; *)
  let open Int64 in
    if n = 1L then
      Printf.printf "%Lu\n" (sub i 2L)
    else
      if (divisible n i) then
        pe003 (div n i) (add i 2L)
      else
        pe003 n (add i 2L) ;;

let main () =
  pe003 600851475143L 3L ;;

let () = main ()
